spec/models/application_spec.rb

```diff
@@ -0,0 +1,36 @@
+RSpec.describe("Application") do
+  include ::Spec::Support::TenantIdentity
+
+  let(:compatible_source)         { Source.create!(:source_type => compatible_source_type, :tenant => tenant, :uid => SecureRandom.uuid, :name => "my-source") }
+  let(:compatible_source_type)    { SourceType.create!(:name => "my-source-type", :product_name => "My Source Type", :vendor => "ACME") }
+
+  let(:incompatible_source)       { Source.create!(:source_type => incompatible_source_type, :tenant => tenant, :uid => SecureRandom.uuid, :name => "not-my-source") }
+  let(:incompatible_source_type)  { SourceType.create!(:name => "not-my-source-type", :product_name => "Not My Source Type", :vendor => "ACME") }
+
+  let(:application_type)  { ApplicationType.create!("name" => "my-application", :supported_source_types => ["my-source-type"]) }
+
+  let(:attributes) do
+    {
+      :application_type_id => application_type.id,
+      :tenant              => tenant
+    }
+  end
+
+  describe "create!" do
+    context "when the application supports the given source type" do
+      it "should return an instance of Application" do
+        expect(
+          Application.create!(attributes.merge(:source_id => compatible_source.id))
+        ).to be_an_instance_of(Application)
+      end
+    end
+
+    context "when the application does not support the given source type" do
+      it "should raise RecordInvalid" do
+        expect do
+          Application.create!(attributes.merge(:source_id => incompatible_source.id))
+        end.to raise_error(ActiveRecord::RecordInvalid, "Validation failed: Source is not compatible with this application type")
+      end
+    end
```

Comment:
It's a little bit weird that we have these contexts, but there's no setup for them that indicates why they matter as contexts.

After looking through the rest of the spec, I understand what's going on, but I think this whole file can actually be refactored a bit to use a `let` for either the source_type name or for the supported source types.

Given the code is checking for inclusion, it seems to make sense that a `let` for the `supported_source_types` should be used instead, and keep the `source_type_name` as an inline "my-source-type". Maybe something like this?:
```ruby
RSpec.describe("Application") do
  include ::Spec::Support::TenantIdentity

  describe "create!" do
    let(:source) { Source.create!(:source_type => source_type, :tenant => tenant, :uid => SecureRandom.uuid, :name => "my-source") }
    let(:source_type) { SourceType.create!(:name => "my-source-type", :product_name => "My Source Type", :vendor => "ACME") }
    let(:application_type)  { ApplicationType.create!("name" => "my-application", :supported_source_types => supported_source_types) }

    subject do
      Application.create!({:application_type_id => application_type.id, :tenant => tenant, :source_id => source.id})
    end

    context "when the application supports the given source type" do
      let(:supported_source_types) { ["my-source-type", "my-other-source-type-not-present-here"] }

      it "returns an instance of Application" do
        expect(subject).to be_an_instance_of(Application)
      end
    end

    context "when the application does not support the given source type" do
      let(:supported_source_types) { ["my-other-source-type-not-present-here"] }

      it "raises a RecordInvalid exception" do
        expect do
          subject
        end.to raise_error(ActiveRecord::RecordInvalid, "Validation failed: Source is not compatible with this application type")
      end
    end
  end
end
```

Reply:
Won't the `let(:application_type)` under the `describe "create!"` be memoized on the first reference, so changing `supported_source_types` won't have any effect on the second reference?

I also think additional test cases are needed for `Application`, that aren't specific to this PR. It would probably make more sense to refactor this spec file after that is fleshed out.

Reply:
@roliveri No, it's lazily loaded for each run-through, and since you would have another lazily loaded `let` for the `supported_source_types`, that gets evaluated at runtime for those specific contexts. I think even if the `let(:application_type)` was `let!(:application_type)` (so, not lazily loaded), due to the fact that you have the `let(:supported_source_types)` within the different contexts means depending on the `it` block that RSpec is currently working on, it should pull in the correct `let`.

Reply:
@eclarizio - Actually, I think it would be clearer if we change the `source` in the set-up for the context, rather than the `supported_source_types`.

We would have the same `subject` set up, and just:
```
let(:source) {  compatible_source }
```
for one context, and:
```
let(:source) {  incompatible_source }
```
for the other.

I prefer this because, it's easier to follow overall, and it more directly illustrates what's being tested - the same application with different sources. As opposed to the same source, and changing the application indirectly.

Reply:
@roliveri That's probably fine, but from the standpoint of the spec, every run through of an `it` block, it *is* a different source already, even though it's referencing the same `let`. I would suggest though to change the context descriptions to reflect that setup, though. It seems odd to say "this is what happens when the application doesn't support X" but the setup doesn't change the application at all.

Like I said, this is kind of subjective, and I just have a lot of opinions built up over the years of working the way I suggested and why I like it the best. I don't want to give the impression you guys *have* to follow what I say 😄

Reply:
@ eclarizio - I think the context descriptions are fine the way they are. True, the application isn't changing between the contexts, but the source is - it shows the application supporting one source but not the other. While it's true, technically it's a different instantiation of the source, but they're instantiated in the same exact way - so they're functionally identical, that's what's significant. The difference between the sources is made clear through the variable names.

I guess it's all how you look at it.

I don't mind subjective suggestions. I don't have much experience writing spec tests, so it's good to get input from those that have. I'll make an update based on my last comments

Reply:
> While it's true, technically it's a different instantiation of the source, but they're instantiated in the same exact way - so they're functionally identical, that's what's significant.

Right, which is why to me, it makes sense to have one source let, and you only adjust the one thing that matters, the source_type_name. Again, nothing wrong with the way you're doing it, it still ends up working, it's mainly in the name of isolating "what sole thing important for this context?". This helps when your specs rely on many different factors, because if you have to keep nesting contexts because you have a ton of different variables, it allows you to see the cyclomatic complexity from the specs where it may have been kind of "hidden" in the code. Then you know you should probably look at pulling some logic out or refactoring into a separate class or something similar.

Reply:
Like I said, it's all in how you look at it. I don't happen agree with this aspect of your argument as it's applied to this specific situation.